src/buildHTML.js

```diff
@@ -17,18 +17,60 @@ var utils = require("./utils");
 
 var makeSpan = buildCommon.makeSpan;
 
+var isSpace = function(node) {
+    return node instanceof domTree.span && node.classes[0] === "mspace";
+};
+
 /**
  * Take a list of nodes, build them in order, and return a list of the built
  * nodes. This function handles the `prev` node correctly, and passes the
- * previous element from the list as the prev of the next element.
+ * previous element from the list as the prev of the next element, ignoring
+ * spaces. documentFragments are flattened into their contents, so the
+ * returned list contains no fragments.
  */
 var buildExpression = function(expression, options, prev) {
+    // Parse expressions into `groups`.
     var groups = [];
     for (var i = 0; i < expression.length; i++) {
         var group = expression[i];
-        groups.push(buildGroup(group, options, prev));
-        prev = group;
+        var output = buildGroup(group, options, prev);
+        if (output instanceof domTree.documentFragment) {
+            Array.prototype.push.apply(groups, output.children);
+        } else {
+            groups.push(output);
+        }
+        if (!isSpace(output)) {
+            prev = group;
+        }
+    }
+    // At this point `groups` consists entirely of `symbolNode`s and `span`s.
+
+    // Explicit spaces (e.g., \;, \,) should be ignored with respect to atom
+    // spacing (e.g., "add thick space between mord and mrel"). Since CSS
+    // adjacency rules implement atom spacing, spaces should be invisible to
+    // CSS. So we shift them into the atoms themselves.
+    for (i = 0; i < groups.length; i++) {
+        if (isSpace(groups[i])) {
+            // Find next non-space group
+            for (var j = i + 1; j < groups.length; j++) {
+                if (!isSpace(groups[j])) {
+                    break;
+                }
+            }
+            // If non-space group exists, ensure that group is a span and tuck
+            // the spaces---nodes [i, j)---into the span as its first children
+            if (j < groups.length) {
+                group = groups[j];
+                if (group instanceof domTree.symbolNode) {
+                    group = groups[j] = makeSpan(group.classes, [group]);
+                }
+                buildCommon.prependChildren(group, groups.splice(i, j - i));
+            } else {
+                i = j;
+            }
+        }
```

Comment:
As for `instanceof domTree.symbolNode`, as the immediately previous comment says,

```
    // At this point `groups` consists entirely of `symbolNode`s and `span`s.
```

because the documentFragments have been spliced in